source/beamline/AMPVControl.cpp

```diff
@@ -672,6 +650,106 @@ void AMPVwStatusAndUnitConversionControl::setUnitConverters(AMAbstractUnitConver
+{
+	disconnect(readPV_, SIGNAL(valueChanged(double)), this, SIGNAL(valueChanged(double)));
+	attemptDouble_ = false;
+	setBinParameters(lowIndex, highIndex);
+	connect(readPV_, SIGNAL(valueChanged()), this, SLOT(onReadPVValueChanged()));
+}
+
+double AMReadOnlyWaveformBinningPVControl::value() const{
+	if(attemptDouble_)
+		return readPV_->binFloatingPointValues(lowIndex_, highIndex_);
+	return readPV_->binIntegerValues(lowIndex_, highIndex_);
+}
+
+void AMReadOnlyWaveformBinningPVControl::setBinParameters(int lowIndex, int highIndex){
+	lowIndex_ = lowIndex;
+	highIndex_ = highIndex;
+}
+
+void AMReadOnlyWaveformBinningPVControl::setAttemptDouble(bool attemptDouble){
+	attemptDouble_ = attemptDouble;
+}
+
+void AMReadOnlyWaveformBinningPVControl::onReadPVValueChanged(){
+	emit valueChanged(value());
+}
+
+/// ========================= implementation of AMWaveformBinningSinglePVControl =================
+AMWaveformBinningSinglePVControl::AMWaveformBinningSinglePVControl(const QString &name, const QString &PVName, int lowIndex, int highIndex, QObject *parent, const QString &description) :
+	AMSinglePVControl(name, PVName, parent)
+{
+	setDescription(description);
+	setAttemptDouble(false);
+	setBinParameters(lowIndex, highIndex);
+
+	// reconnect the valueChanged signal for readPV_
+	disconnect(readPV_, SIGNAL(valueChanged(double)), this, SIGNAL(valueChanged(double)));
+	connect(readPV_, SIGNAL(valueChanged()), this, SLOT(onReadPVValueChanged()));
+}
+
+double AMWaveformBinningSinglePVControl::value() const{
+	if(attemptDouble_)
+		return readPV_->binFloatingPointValues(lowIndex_, highIndex_);
+	return readPV_->binIntegerValues(lowIndex_, highIndex_);
+}
+
+bool AMWaveformBinningSinglePVControl::values(int size, int *outputValues) const{
+	if(attemptDouble_)
+		return false;
+	QVector<int> integerValues = readPV_->lastIntegerValues();
+	if(integerValues.count() < size)
+		return false;
+	memcpy(outputValues, integerValues.constData(), size);
+	return true;
+}
+
+bool AMWaveformBinningSinglePVControl::values(int size, double *outputValues) const{
+	if(!attemptDouble_)
+		return false;
+	QVector<double> doubleValues = readPV_->lastFloatingPointValues();
+	if(doubleValues.count() < size)
+		return false;
+	memcpy(outputValues, doubleValues.constData(), size);
+	return true;
+}
+
+void AMWaveformBinningSinglePVControl::setBinParameters(int lowIndex, int highIndex){
+	lowIndex_ = lowIndex;
+	highIndex_ = highIndex;
+}
+
+void AMWaveformBinningSinglePVControl::setAttemptDouble(bool attemptDouble){
+	attemptDouble_ = attemptDouble;
+}
+
+void AMWaveformBinningSinglePVControl::setValues(const QVector<int> values){
+//	qDebug() << "Attempting QVector<int>";
+	if(attemptDouble_)
+		return;
+
+//	qDebug() << "Doing QVector<int> on " << writePV_->pvName();
+
+	writePV_->setValues(((dbr_long_t*)(values.data())), values.count());
+}
+
+void AMWaveformBinningSinglePVControl::setValues(const QVector<double> values){
```

Comment:
You should make sure that these vectors are `const QVector<double> &values)` to guarantee that implicit sharing is being used.